toolchain/lex/lex.cpp

```diff
@@ -1044,41 +1130,67 @@ auto Lexer::LexOneCharSymbolToken(llvm::StringRef source_text, TokenKind kind,
       << "' instead of the spelling '" << kind.fixed_spelling()
       << "' of the incoming token kind '" << kind << "'";
 
-  TokenIndex token = buffer_.AddToken(
-      {.kind = kind, .byte_offset = static_cast<int32_t>(position)});
+  TokenIndex token = LexToken(kind, position);
   ++position;
   return token;
 }
 
 auto Lexer::LexOpeningSymbolToken(llvm::StringRef source_text, TokenKind kind,
                                   ssize_t& position) -> LexResult {
-  TokenIndex token = LexOneCharSymbolToken(source_text, kind, position);
+  CARBON_DCHECK(kind.is_opening_symbol());
+  CARBON_DCHECK(kind.fixed_spelling().size() == 1);
+  CARBON_DCHECK(source_text[position] == kind.fixed_spelling().front())
+      << "Source text starts with '" << source_text[position]
+      << "' instead of the spelling '" << kind.fixed_spelling()
+      << "' of the incoming token kind '" << kind << "'";
+
+  int32_t byte_offset = position;
+  ++position;
+
+  // Lex the opening symbol with a zero closing index. We'll update this
+  // later if we successfully match it to a closing symbol.
+  TokenIndex token =
+      AddLexedToken(TokenInfo(kind, has_leading_space_, 0, byte_offset));
```

Comment:
What made you choose to add the `Lexer::LexToken` overloads instead of having these provided as constructors on `TokenInfo` and writing `AddLexedToken(TokenInfo(...))`? This particularly feels odd to me since, in some spots, code is doing `AddLexedToken(TokenInfo(...))` directly (e.g., `LexOpeningSymbolToken`). Why don't the calls align?

To note there are a few paths above -- I think you are providing multiple ways to handle the differences:

- Overloads of `LexToken`, which pass the appropriate payload setup to `TokenInfo`.
- Two `TokenInfo` constructors, one if there's no payload.
- Calling `AddLexedToken` directly with an explicit `0` for no payload.

Particularly with the last, maybe you were iterating and missed it? I'd suggest choosing between:

- One `TokenInfo` constructor, have `LexToken` choose the payload (including `0`).
- Multiple `TokenInfo` constructors, remove `LexToken`, only do `AddLexedToken(TokenInfo(...))`

I have a slight leaning towards the latter because you couldn't have code "accidentally" do `AddLexedToken(TokenInfo(...))` itself (it would be the right way to do it), but `LexToken` is less typing.

Reply:
So, the real reason I wanted to have `LexToken` is to try to have one place where `has_leading_space_` needs to be remembered and passed in to the token construction, as that is literally the same for every token (unlike byte offset for example).

The secondary thing I like is that it concentrates more of the logic in the lexer rather than in the tokenized buffer that is more about *consuming* tokens.

But I agree that the complex overloads aren't helping here. *And*, as you noticed here, I failed to use them consistently. I've tried to simplify this down to just a payload overload and a non-payload overload. There is only one caller for most of these, and so I think the type checking we're getting isn't really that valuable.

I've also switched this code to use them correctly rather than directly constructing things here. This part of the PR was a bit rushed, sorry about that. Should be much cleaner now.

WDYT?

Reply:
This looks good, thanks!